### ISSUE TITLE
Small payment from lnd node to c-lightning node results in "Internal error"

### ISSUE THREAD

**Issue by grubles** [lightningnetwork/lnd#680]
I have a channel open between my lnd node and Blockstream's c-lightning node. I received a small invoice of 1 satoshi. I attempted to pay it but now receive the following output/errors from my lnd. It seems an Internal error happens on the c-lightning end, disconnects my node, reconnects, and repeats the error:
(What I think are the relevant logs. I can add more if asked. Also, this issue is duplicated here as it is unclear to me where the issue lies: https://github.com/ElementsProject/lightning/issues/814)

```
[INF] HSWC: Attempting to re-resynchronize ChannelPoint(83befa3d40045afc7c7537247911506a7598763714328215ec00c5f556884d45:1)
2018-01-27 13:29:27.260 [DBG] PEER: Sending ChannelAnnouncement(chain_hash=000000000019d6689c085ae165831e934ff763ae46a2a6c172b3f1b60a8ce26f, short_chan_id=552569464202199041) to 104.198.32.198:9735
2018-01-27 13:29:27.260 [DBG] PEER: Received Error(chan_id=0000000000000000000000000000000000000000000000000000000000000000, err=Internal error) from 104.198.32.198:9735
2018-01-27 13:29:27.260 [DBG] PEER: Sending ChannelUpdate(chain_hash=000000000019d6689c085ae165831e934ff763ae46a2a6c172b3f1b60a8ce26f, short_chan_id=552569464202199041, flag=0, update_time=2018-01-27 12:07:35 -0500 EST) to 104.198.32.198:9735
2018-01-27 13:29:27.260 [INF] PEER: unable to read message from 104.198.32.198:9735: EOF
2018-01-27 13:29:27.261 [DBG] PEER: Sending NodeAnnouncement(node=03bd3466efd4a7306b539e2314e69efc6b1eaee29734fcedd78cf81b1dde9fedf8, update_time=2018-01-25 20:10:18 -0500 EST) to 104.198.32.198:9735
2018-01-27 13:29:27.261 [DBG] SRVR: Peer 104.198.32.198:9735 has been disconnected
2018-01-27 13:29:27.261 [DBG] PEER: Peer shutting down, could not enqueue msg.
2018-01-27 13:29:27.261 [DBG] PEER: Peer shutting down, could not enqueue msg.
2018-01-27 13:29:27.261 [WRN] FNDG: Received error for non-existent funding flow: (*lnwire.Error)(0xc422a92240)({
 ChanID: (lnwire.ChannelID) (len=32 cap=32) 0000000000000000000000000000000000000000000000000000000000000000,
 Data: (lnwire.ErrorData) (len=14 cap=14) {
  00000000  49 6e 74 65 72 6e 61 6c  20 65 72 72 6f 72        |Internal error|
 }
})

2018-01-27 13:29:27.261 [DBG] FNDG: Cancelling all reservations for peer 02f6725f9c1c40333b67faea92fd211c183050f28df32cac3f9d69685fe9665432
2018-01-27 13:29:27.261 [DBG] FNDG: No active reservations for node: 02f6725f9c1c40333b67faea92fd211c183050f28df32cac3f9d69685fe9665432
2018-01-27 13:29:27.261 [DBG] PEER: Peer shutting down, could not enqueue msg.
2018-01-27 13:29:27.261 [INF] HSWC: Removing channel link with ChannelID(454d8856f5c500ec15823214377698756a5011792437757cfc5a04403dfabe82)
2018-01-27 13:29:27.261 [INF] HSWC: ChannelLink(83befa3d40045afc7c7537247911506a7598763714328215ec00c5f556884d45:1) is stopping
2018-01-27 13:29:27.261 [DBG] PEER: Peer shutting down, could not enqueue msg.
2018-01-27 13:29:27.261 [DBG] PEER: Peer shutting down, could not enqueue msg.
2018-01-27 13:29:27.261 [DBG] PEER: Peer shutting down, could not enqueue msg.
2018-01-27 13:29:27.261 [DBG] PEER: Peer shutting down, could not enqueue msg.
2018-01-27 13:29:27.261 [DBG] PEER: Peer shutting down, could not enqueue msg.
```

**Comment by Roasbeef** [comment]
This _may_ be related to #547

**Comment by halseth** [comment]
I haven't been able to reproduce this (as you mention, an internal error probably happened with `c-lightning`), but looking at the logs, this is my take on the issue:

* `c-lightning` experiences an internal error and sends a "connection wide error", prompting us to fail all channels with the peer. It then disconnects.
* since `lnd` currently doesn't interpret these errors as connection wide, we would just disconnect the peer, and the error would repeat when the node connects again (if I understand the initial description correctly, the same logs would repeat over and over? Would be great if you could post a bigger log chunk of you still have them @grubles)

This should be partly fixed by #790, where we start recognizing these connection wide errors. However, we still don't mark the channel as "failed" in any way (currently we just disconnect the peer), which would lead to this happening again on next connection to the peer. Gracefully failing channels in case of errors is in the works.

The spec mentions that on attempted channel re-sync, the receiving node should resend the `Error` message for failed channels, which is probably why `c-lightning` node would resend the `Error` on connection. Two possible solutions to this I see is 
1) Mark the channel as failed, such that we don't attempt to re-sync it at start up.
2) For nodes that are connected and then quickly disconnected, we should have an exponential backoff for reconnection.

I think 1) is the most correct one (as the other node is not really misbehaving), while 2) could also help for other cases of misbehaving nodes.

**Comment by Roasbeef** [comment]
Thanks for investigating @halseth! 

Closing.
